File: dotnet/cafdemalihapa/Applications/DirectoryManagement/Opening.cs
using cafdemalihapa.Applications.DirectoryManagement.DirectoryOpenStrategies;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.AutomationsDirectory;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.AutomationsDirectory.CommandsDirectory;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.AutomationsDirectory.EnvironmentVariablesFiles;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.AutomationsDirectory.OperationsDirectory;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.BackEnd;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.BackEnd.BackEndPrimaryDirectory;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.BackEnd.BackEndSecondaryDirectory;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.Calls;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.DataDirectory;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.FrontEndDirectory;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.FrontEndDirectory.FrontEndGuestDirectory;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.FrontEndDirectory.FrontEndHostDirectory;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.NotesAndMessages;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.Tools;
using cafdemalihapa.Directories.HostingDirectory.FeatureDirectory.WebLinks;
using cafdemalihapa.Names;
using Microsoft.Extensions.Logging;

namespace cafdemalihapa.Applications.DirectoryManagement
{
    public class Opening
    (
        ILogger<Opening> logger,
        FeatureDirectory featureDirectory,
        AutomationsDirectory automationsDirectory,
        CommandsDirectory commandsDirectory,
        OperationsDirectory operationsDirectory,
        DataDirectory dataDirectory,
        ToolsDirectory toolsDirectory,
        BackEndDirectory backEndDirectory,
        BackEndPrimaryDirectory backEndPrimaryDirectory,
        BackEndSecondaryDirectory backEndSecondaryDirectory,
        CallsDirectory callsDirectory,
        FrontEndDirectory frontEndDirectory,
        FrontEndHostDirectory frontEndHostDirectory,
        FrontEndGuestDirectory frontEndGuestDirectory,
        NotesAndMessagesDirectory notesAndMessagesDirectory,
        WebLinksDirectory webLinksDirectory,
        EnvironmentVariablesFilesDirectory environmentVariablesFilesDirectory,
        DirectoryToBeOpen directoryToBeOpen,
        IEnumerable<IDirectoryOpenStrategy> directoryOpenStrategies
    )
    {
        public void OpenDirectoryToBeOpen()
        {
            if (!Commands.Get().IsDirectoryToBeOpen())
            {
                return;
            }
            var directoryToBeOpenPath = directoryToBeOpen.GetPath();
            OpenDirectories(directoryToBeOpenPath);
        }

        public void Open()
        {
            if (!Commands.Get().IsOpen())
            {
                return;
            }

            var environmentVariablesSourceDictionary = environmentVariablesFilesDirectory.PairUp();

            var directoriesToOpen = new Dictionary<string, string>
            {
                { "IS_OPENING_FEATURE_SELF_ADDRESS", featureDirectory.GetPath() },
                { "IS_OPENING_AUTOMATIONS_DIRECTORY", automationsDirectory.GetPath() },
                { "IS_OPENING_COMMANDS_DIRECTORY", commandsDirectory.GetPath() },
                { "IS_OPENING_ENVIRONMENT_VARIABLES_FILES_DIRECTORY", environmentVariablesFilesDirectory.GetPath() },
                { "IS_OPENING_OPERATIONS_DIRECTORY", operationsDirectory.GetPath() },
                { "IS_OPENING_BEND_ADDRESS", backEndDirectory.GetPath() },
                { "IS_OPENING_BEND_PRIMARY_ADDRESS", backEndPrimaryDirectory.GetPath() },
                { "IS_OPENING_BEND_SECONDARY_ADDRESS", backEndSecondaryDirectory.GetPath() },
                { "IS_OPENING_DATA_DIRECTORY", dataDirectory.GetPath() },
                { "IS_OPENING_FEND_ADDRESS", frontEndDirectory.GetPath() },
                { "IS_OPENING_FEND_HOST_ADDRESS", frontEndHostDirectory.GetPath() },
                { "IS_OPENING_FEND_GUEST_ADDRESS", frontEndGuestDirectory.GetPath() },
                { "IS_OPENING_FEND_PRIMARY_ADDRESS", frontEndHostDirectory.GetPath() },
                { "IS_OPENING_FEND_SECONDARY_ADDRESS", frontEndGuestDirectory.GetPath() },
                { "IS_OPENING_CALLS_ADDRESS", callsDirectory.GetPath() },
                { "IS_OPENING_TOOLS_ADDRESS", toolsDirectory.GetPath() },
                { "IS_OPENING_NOTES_MESSAGES_ADDRESS", notesAndMessagesDirectory.GetPath() },
                { "IS_OPENING_WEB_LINKS_ADDRESS", webLinksDirectory.GetPath() },
            };

            foreach (var (key, path) in directoriesToOpen)
            {
                if (environmentVariablesSourceDictionary.TryGetValue(key, out string? value))
                {
                    if (bool.TryParse(value, out bool shouldOpen) && shouldOpen)
                    {
                        OpenDirectories(path);
                        Thread.Sleep(700);
                    }
                }
            }
        }

        public void OpenDirectories(string path)
        {
            var strategy = directoryOpenStrategies.FirstOrDefault(s => s.CanHandle());

            if (strategy is null)
            {
                logger.LogWarning("Opening: no directory-open strategy supports the current operating system.");
                return;
            }

            strategy.Open(path);
        }
    }
}

File: dotnet/cafdemvers/zzq-cafdemwimu/cafdemwimu.console/Applications/DirectoryManagement/Opening.fs
namespace cafdemwimu.console.Applications.DirectoryManagement

open System
open System.Threading
open System.Collections.Generic
open Microsoft.Extensions.Logging
open cafdemwimu.console.Applications
open cafdemwimu.console.Applications.DirectoryManagement.DirectoryOpenStrategies
open cafdemwimu.console.Names
open cafdemwimu.console.Directories.Hosting.Feature
open cafdemwimu.console.Directories.Hosting.Feature.Automations
open cafdemwimu.console.Directories.Hosting.Feature.Automations.Commands
open cafdemwimu.console.Directories.Hosting.Feature.Automations.EnvironmentVariablesFiles
open cafdemwimu.console.Directories.Hosting.Feature.Automations.Operations
open cafdemwimu.console.Directories.Hosting.Feature.BackEnd
open cafdemwimu.console.Directories.Hosting.Feature.BackEnd.BackEndPrimary
open cafdemwimu.console.Directories.Hosting.Feature.BackEnd.BackEndSecondary
open cafdemwimu.console.Directories.Hosting.Feature.Calls
open cafdemwimu.console.Directories.Hosting.Feature.Data
open cafdemwimu.console.Directories.Hosting.Feature.FrontEnd
open cafdemwimu.console.Directories.Hosting.Feature.FrontEnd.FrontEndPrimary
open cafdemwimu.console.Directories.Hosting.Feature.FrontEnd.FrontEndSecondary
open cafdemwimu.console.Directories.Hosting.Feature.NotesAndMessages
open cafdemwimu.console.Directories.Hosting.Feature.Tools
open cafdemwimu.console.Directories.Hosting.Feature.WebLinks

type Opening
    (
        logger: ILogger<Opening>,
        automationsDirectory: AutomationsDirectory,
        commandsDirectory: CommandsDirectory,
        operationsDirectory: OperationsDirectory,
        dataDirectory: DataDirectory,
        toolsDirectory: ToolsDirectory,
        backEndPrimaryDirectory: BackEndPrimaryDirectory,
        backEndSecondaryDirectory: BackEndSecondaryDirectory,
        callsDirectory: CallsDirectory,
        frontEndDirectory: FrontEndDirectory,
        frontEndPrimaryDirectory: FrontEndPrimaryDirectory,
        frontEndSecondaryDirectory: FrontEndSecondaryDirectory,
        notesAndMessagesDirectory: NotesAndMessagesDirectory,
        webLinksDirectory: WebLinksDirectory,
        environmentVariablesFilesDirectory: EnvironmentVariablesFilesDirectory,
        directoryToBeOpen: DirectoryToBeOpen,
        directoryOpenStrategies: IEnumerable<IDirectoryOpenStrategy>
    ) =
    member this.OpenDirectoryToBeOpen() =
        if Commands.Get().IsDirectoryToBeOpen() then
            let directoryToBeOpenPath = directoryToBeOpen.GetPath()
            this.OpenDirectories(directoryToBeOpenPath)

    member this.Open() =
        if Commands.Get().IsOpen() then
            let environmentVariablesSourceDictionary = environmentVariablesFilesDirectory.PairUp()

            let directoriesToOpen = Dictionary<string, string>()
            directoriesToOpen.Add("IS_OPENING_FEATURE_SELF_ADDRESS", FeatureDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_AUTOMATIONS_DIRECTORY", automationsDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_COMMANDS_DIRECTORY", commandsDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_ENVIRONMENT_VARIABLES_FILES_DIRECTORY", environmentVariablesFilesDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_OPERATIONS_DIRECTORY", operationsDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_BEND_ADDRESS", BackEndDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_BEND_PRIMARY_ADDRESS", backEndPrimaryDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_BEND_SECONDARY_ADDRESS", backEndSecondaryDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_DATA_DIRECTORY", dataDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_FEND_ADDRESS", frontEndDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_FEND_PRIMARY_ADDRESS", frontEndPrimaryDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_FEND_SECONDARY_ADDRESS", frontEndSecondaryDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_CALLS_ADDRESS", callsDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_TOOLS_ADDRESS", toolsDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_NOTES_MESSAGES_ADDRESS", notesAndMessagesDirectory.GetPath())
            directoriesToOpen.Add("IS_OPENING_WEB_LINKS_ADDRESS", webLinksDirectory.GetPath())

            for entry in directoriesToOpen do
                match environmentVariablesSourceDictionary.TryGetValue entry.Key with
                | true, value ->
                    match Boolean.TryParse(value) with
                    | true, shouldOpen when shouldOpen ->
                        this.OpenDirectories(entry.Value)
                        Thread.Sleep(700)
                    | _ -> ()
                | _ -> ()

    member _.OpenDirectories(path: string) =
        let strategy = directoryOpenStrategies |> Seq.tryFind (fun s -> s.CanHandle())

        match strategy with
        | None ->
            logger.LogWarning("Opening: no directory-open strategy supports the current operating system.")
        | Some strategy ->
            strategy.Open(path)
